sklearn/utils/estimator_checks.py

```diff
@@ -2565,6 +2567,60 @@ def check_classifiers_regression_target(name, estimator_orig):
         assert_raises_regex(ValueError, msg, e.fit, X, y)
 
 
+@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+def check_estimator_sparse_dense(name, estimator_orig):
+    rng = np.random.RandomState(52)
+    estimator = clone(estimator_orig)
+    estimator_sp = clone(estimator_orig)
+    tags = _safe_tags(estimator_orig)
+    centers = 2 if tags["binary_only"] else None
+    X, y = make_blobs(random_state=rng, cluster_std=0.5, centers=centers)
+    # for put some points to zero to have a little bit of sparsity
+    # TODO: update this with enforce_estimator_tags_X when #14705 is merged
+    if tags["requires_positive_X"]:
+        X -= np.min(X) - 1.
+    X_csr = sparse.csr_matrix(X)
+    if tags["multioutput"] or tags["multioutput_only"]:
+        y = y[:, np.newaxis]
+
+    for sparse_format in ['csr', 'csc', 'dok', 'lil', 'coo', 'dia', 'bsr']:
+        X_sp = X_csr.asformat(sparse_format)
+        if name in ['Scaler', 'StandardScaler']:
+            estimator.set_params(with_mean=False)
+            estimator_sp.set_params(with_mean=False)
+        if name in ['RobustScaler']:
+            estimator.set_params(with_centering=False)
+            estimator_sp.set_params(with_centering=False)
+
+        set_random_state(estimator)
+        set_random_state(estimator_sp)
+        X_converted = pairwise_estimator_convert_X(X, estimator)
+        X_sp_converted = pairwise_estimator_convert_X(X_sp, estimator)
+        try:
+            estimator_sp.fit(X_sp_converted, y)
+            estimator.fit(X_converted, y)
+            if hasattr(estimator, "predict"):
+                pred = estimator.predict(X_converted)
+                pred_sp = estimator_sp.predict(X_sp_converted)
+                if not (isinstance(estimator, BaseSGD) or  # SGD is too random
+                        getattr(estimator, 'kernel', None) == 'precomputed' or
```

Comment:
Removing these lines, it works for pairwise estimators (maybe because we use `_pairwise_estimator_convert_X` before ?)
But it does not work for `BaseSGD`, which seems to be non deterministic, I tried to look for an issue about that but it seems that the flag `not_deterministic` is not given to `BaseSGD`, maybe it should ?

Reply:
`BaseSGD` should be deterministic (assuming you fix the `random_state`). I believe `not_deterministic` is better suited for other uncontrollable sources of non-determinism (e.g. slight numerical variation caused by different ordering of arithmetic operations caused by thread-based parallelism that causes ties in cluster assignment to be broken in different ways when running k-means clustering for instance).